operator/controllers/operator/eventexporter_controller.go

```diff
@@ -0,0 +1,218 @@
+}
+
+// +kubebuilder:rbac:groups=operator.skywalking.apache.org,resources=eventexporters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=operator.skywalking.apache.org,resources=eventexporters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=operator.skywalking.apache.org,resources=eventexporters/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+func (r *EventExporterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := runtimelog.FromContext(ctx)
+	log.Info(fmt.Sprintf("===============eventexporter reconcile started (ns: %s, name: %s)===============", req.Namespace, req.Name))
+
+	eventExporter := operatorv1alpha1.EventExporter{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &eventExporter); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	newConfigMapName := configMapName(&eventExporter)
+	if _, err := r.overlayData(ctx, log, &eventExporter, newConfigMapName); err != nil {
+		log.Error(err, "failed to delete eventexporter's configMap")
+		return ctrl.Result{}, err
+	}
+
+	ff, err := r.FileRepo.GetFilesRecursive("templates")
+	if err != nil {
+		log.Error(err, "failed to load resource templates")
+		return ctrl.Result{}, err
+	}
+
+	app := kubernetes.Application{
+		Client:   r.Client,
+		FileRepo: r.FileRepo,
+		CR:       &eventExporter,
+		GVK:      operatorv1alpha1.GroupVersion.WithKind("EventExporter"),
+		Recorder: r.Recorder,
+		TmplFunc: template.FuncMap{
+			"configMapName": func() string { return newConfigMapName },
+		},
+	}
+
+	if err := app.ApplyAll(ctx, ff, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.checkState(ctx, log, &eventExporter); err != nil {
+		l.Error(err, "failed to check sub resources state")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{RequeueAfter: schedDuration}, nil
+
+}
+
+func (r *EventExporterReconciler) overlayData(ctx context.Context, log logr.Logger,
+	eventExporter *operatorv1alpha1.EventExporter, newConfigMapName string) (changed bool, err error) {
+
+	oldConfigMapName := eventExporter.Status.ConfigMapName
+	if oldConfigMapName == newConfigMapName {
+		log.Info("eventexporter configuration keeps the same as before")
+		return false, nil
+	}
+
+	configmap := core.ConfigMap{}
+	err = r.Client.Get(ctx, client.ObjectKey{Namespace: eventExporter.Namespace, Name: oldConfigMapName}, &configmap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to get the eventexporter's configmap")
+		return true, err
+	}
+
+	if !apierrors.IsNotFound(err) {
+		if err = r.Client.Delete(ctx, &configmap); err != nil {
+			log.Error(err, "failed to delete eventexporter's configmap")
+			return true, err
```

Comment:
We'd better to create the new CM first, then delete the old one. So the returned `changed: true` makes sense.

Reply:
Thx, updated!